docs/architecture/psa-keystore-design.md

```diff
@@ -0,0 +1,197 @@
+
+#### Length of the volatile key slices
+
+The volatile key slices have exponentially increasing length: each slice is twice as long as the previous one. Thus if the length of slice 0 is `B` and there are `N` slices, then there are `B * (2^N - 1)` slots.
+
+As of Mbed TLS 3.6.1, the maximum number of volatile key slots is less than the theoretical maximum of 2^30 - 2^16 (0x10000000..0x7ffeffff, the largest range of key identifiers reserved for the PSA Crypto implementation that does not overlap the range for built-in keys). The reason is that we limit the slot index to 2^25-1 so that the [encoding of volatile key identifiers](#volatile-key-identifiers-in-the-dynamic-key-store) has 25 bits for the slot index.
+
+When `MBEDTLS_TEST_HOOKS` is enabled, the length of key slices can be overridden. We use this in tests that need to fill the key store.
+
+#### Free list
+
+Each volatile key slice has a **free list**. This is a linked list of all the slots in the slice that are free. The global data contains a static array of free list heads, i.e. the index of a free slot in the slice. Each free slot contains the index of the next free slot in that slice's free list. The last element of the linked list (or the head if the list is empty) contains an index that is larger than the length of the slice,
+
+As a small optimization, a free slot does not actually contain the index of the next slot, but the index of the next free slot on the list _relative to the next slot in the array_. This way, a slice freshly obtained from `calloc` has all of its slots in the free list in order. The absolute index of the next slot after slot `i` in the free list is `i + slice[i].next_free_relative_to_next`.
+
+#### Dynamic key slot allocation
+
+To create a volatile key, `psa_reserve_free_key_slot()` searches the free lists of each allocated slice until it finds a slice that is not full. If all allocated slices are full, the code allocates a new slice at the lowest possible slice index. If all possible slices are already allocated and full, the key creation fails.
+
+The newly allocated slot is removed from the slice's free list.
+
+We only allocate a slice of size `B * 2^k` if there are already `B * (2^k - 1)` occupied slots. Thus the memory overhead is at most `B` slots plus the number of occupied slots, i.e. the memory consumption for slots is at most than twice the required memory plus a small constant overhead.
+
+#### Dynamic key slot deallocation
+
+When destroying a volatile key, `psa_wipe_key_slot()` calls `psa_free_key_slot()`. This function adds the newly freed slot to the head of the free list.
+
+##### Future improvement: slice deallocation
+
+As of Mbed TLS 3.6.1, `psa_free_key_slot()` does not deallocate slices. Thus the memory consumption for slots never decreases (except when the PSA crypto subsystem is deinitialized). Freeing key slices intelligently would be a desirable improvement.
```

Comment:
Does this constitute something like a memory leak? I suppose that we never exceed a fixed value, so it isn't cumulative like most problematic memory leaks, but it still could be a problem if we have a program that spawns many connections once briefly for some purpose and then performs some other tasks.

Should we warn our users about this? Should we advise them to deinit and reinit PSA every once in a while if they have this issue?

Could we provide some API to manually call that deletes empty key slots for users who care about this?

Reply:
It's a leak in the sense that an application won't get back all of its memory if it stops using many keys, but not in the sense that the memory is completely lost. I agree that it warrants a warning in the documentation.

We can't free empty slots, only empty slices. If a slice contains a slot in use, it has to stay. Compaction is effectively impossible without a major redesign of our threading approach, and that in turns will require better primitives than a mutex, so it's a post-4.0 thing.

Reply:
Sorry, I meant could we provide a function to free all of the fully empty slices, starting from the largest and ending before the first that has a key slot allocated.

That would mean at least that users who have bursty many-connection TLS 1.3 applications could do something like:
```c
/* Spawn lots of TLS connections to avoid blocking */
for (int i = 0; i < num_requests_to_make; i++) {
    threads[i % num_workers] = spawn_new_tls_connection(requests_to_make[i]);
}

/* Reduce memory usage by freeing key slices */
mbedtls_psa_cleanup_empty_slices();

/* Do other memory-intensive work */
```

Maybe we wouldn't want to tie ourselves to maintaining something like this, but it seems like a problem to have no possible way of ever deallocating the slices. Wdyt?

Reply:
I maintain that this is too much for 3.6.1. This is potentially 3.6.2 material, depending on whether users complain. But I don't want to make it a user interface: most users wouldn't know when to call the function. This is best done as an automatic decision, and one that can be tuned without it affecting the functional behavior of applications.